### PR TITLE
process validation: allow tasks without outfilegrp

### LINKED ISSUE #295: ocrd process: allow non-output tasks
It is agreed there are processors which create no actual output file groups, but are rather run to produce logs or model files. Please support this in `ocrd process`, which currently raises an exception about that.

### PR DESCRIPTION
**Pull request by bertsky**
Fixes #295

### CODE REVIEW

**Comment by bertsky** [comment]
Obviously I forgot to adapt the validator test. So do we skip the offending `test_parse_no_out` altogether, or is there anything we could test instead?

**Comment by bertsky** [comment]
Maybe the validator can be made sensitive of whether the task (as per `ocrd-tool`) should actually have an output file group or not?

File: ocrd/ocrd/task_sequence.py [hunk 1]
```diff
@@ -38,14 +39,18 @@ def __init__(self, executable, input_file_grps, output_file_grps, parameter_path
...
-        if self.parameter_path and not os.access(self.parameter_path, os.R_OK):
-            raise Exception("Parameter file not readable: %s" % self.parameter_path)
-        if not self.input_file_grps:
-            raise Exception("Task must have input file group")
-        if not self.output_file_grps:
-            raise Exception("Task must have output file group")
         if not which(self.executable):
             raise Exception("Executable not found in PATH: %s" % self.executable)
+        result = run([self.executable, '--dump-json'], stdout=PIPE, check=True, universal_newlines=True)
+        ocrd_tool_json = json.loads(result.stdout)
```
Comment (bertsky) [inline ocrd/ocrd/task_sequence.py]: Nice! I have to remember that.

File: ocrd/ocrd/task_sequence.py [hunk 2]
```diff
@@ -38,14 +39,18 @@ def __init__(self, executable, input_file_grps, output_file_grps, parameter_path
...
+        # TODO check for required parameters in ocrd_tool
+        if self.parameter_path:
+            parse_json_string_or_file(self.parameter_path)
```
Comment (bertsky) [inline ocrd/ocrd/task_sequence.py]: Yes, this finally opens the possibility of also checking that:
- the parameter file does not contain keys that are not in tool.json
- the parameter file specifies a value for every parameter that does not have a default in tool.json (#246)

File: tests/test_utils.py [hunk 3]
```diff
@@ -168,5 +172,27 @@ def test_nth_url_segment(self):
...
+        self.assertEqual(parse_json_string_or_file('{"foo": 32}'), {'foo': 32})
+
+    def test_parameter_file(self):
+        with TemporaryDirectory() as tempdir:
+            paramfile = str(Path(tempdir, '{}')) # XXX yes, the file is called '{}'
+            with open(paramfile, 'w') as f:
+                f.write('{"bar": 42}')
+            self.assertEqual(parse_json_string_or_file(paramfile), {'bar': 42})
+            with pushd_popd(tempdir):
+                self.assertEqual(parse_json_string_or_file(), {'bar': 42})
```
[the remark this answers is not in the document]
Comment (kba) [reply tests/test_utils.py]: Added as a comment.

File: tests/test_utils.py [hunk 4]
```diff
@@ -168,5 +172,27 @@ def test_nth_url_segment(self):
         self.assertEqual(nth_url_segment('/path/to/foo#frag', n=-2), 'to')
         self.assertEqual(nth_url_segment('https://server/foo?xyz=zyx'), 'foo')
 
+    def test_parse_json_string_or_file(self):
+        self.assertEqual(parse_json_string_or_file(), {})
+        self.assertEqual(parse_json_string_or_file('{}'), {})
+        self.assertEqual(parse_json_string_or_file('{"foo": 32}'), {'foo': 32})
+        self.assertEqual(parse_json_string_or_file('{"foo": 32}'), {'foo': 32})
```
[the remark this answers is not in the document]
Comment (kba) [reply tests/test_utils.py]: artifact from refactoring, should ensure that valid JSON as a filename is parsed as a literal if the file doesn't exist. I've fixed the test.

Funny how permissive POSIX filenames can be.

**Comment by kba** [comment]
> we also need to adapt spec accordingly

Already adapted, `output_file_grp` is optional in ocrd-tool.json and CLI spec lists parameter as optional.

File: ocrd/ocrd/task_sequence.py (hunk 2 above)
Comment (kba) [reply ocrd/ocrd/task_sequence.py]: I'll add these in subsequent PRs, #251 #246 are not forgotten.

**Pull request merged**